tools/state-viewer/src/commands.rs

```diff
@@ -649,3 +660,118 @@ pub fn format_hash(h: CryptoHash) -> String {
 pub fn chunk_mask_to_str(mask: &[bool]) -> String {
     mask.iter().map(|f| if *f { '.' } else { 'X' }).collect()
 }
+
+fn filter_hashes<T: Clone, F: Fn(&T) -> CryptoHash>(
+    items: &Vec<T>,
+    item_hash: F,
+    hashes: Option<Vec<CryptoHash>>,
+) -> Result<Vec<T>, CryptoHash> {
+    match hashes {
+        Some(hashes) => {
+            let mut hashes_seen =
+                hashes.iter().map(|h| (h.clone(), false)).collect::<HashMap<_, _>>();
+
+            let filtered = items
+                .iter()
+                .filter_map(|x| {
+                    let hash = item_hash(x);
+                    match hashes_seen.entry(hash) {
+                        Entry::Occupied(mut e) => {
+                            *e.get_mut() = true;
+                            Some(x.clone())
+                        }
+                        Entry::Vacant(_) => None,
+                    }
+                })
+                .collect();
+            for (hash, seen) in hashes_seen.iter() {
+                if !*seen {
+                    return Err(*hash);
+                }
+            }
+            Ok(filtered)
+        }
+        None => Ok(items.clone()),
+    }
+}
+
+fn filter_txs(
+    chunk: &ShardChunk,
+    tx_hashes: Option<Vec<CryptoHash>>,
+) -> anyhow::Result<Vec<SignedTransaction>> {
+    filter_hashes(chunk.transactions(), |tx| tx.get_hash(), tx_hashes).map_err(|hash| {
+        anyhow!("transaction with hash {} not found in chunk {:?}", hash, chunk.chunk_hash())
+    })
+}
+
+fn filter_receipts(
+    chunk: &ShardChunk,
+    receipt_hashes: Option<Vec<CryptoHash>>,
+) -> anyhow::Result<Vec<Receipt>> {
+    filter_hashes(chunk.receipts(), |r| r.receipt_id, receipt_hashes).map_err(|hash| {
+        anyhow!("receipt with ID {} not found in chunk {:?}", hash, chunk.chunk_hash())
+    })
+}
+
+pub(crate) fn apply_chunk(
+    home_dir: &Path,
+    near_config: NearConfig,
+    store: Store,
+    chunk_hash: ChunkHash,
+    tx_hashes: Option<Vec<CryptoHash>>,
+    receipt_hashes: Option<Vec<CryptoHash>>,
+) -> anyhow::Result<()> {
+    let mut chain_store = ChainStore::new(store.clone(), near_config.genesis.config.genesis_height);
+    let runtime = Arc::new(NightshadeRuntime::with_config(
+        home_dir,
+        store,
+        &near_config,
+        None,
+        near_config.client_config.max_gas_burnt_view,
+    ));
+    let chunk = chain_store.get_chunk(&chunk_hash)?;
+    let chunk_header = chunk.cloned_header();
+
+    let prev_block_hash = chunk_header.prev_block_hash();
+    let shard_id = chunk.shard_id();
+    let prev_state_root = chunk.prev_state_root();
+
+    let transactions = filter_txs(chunk, tx_hashes)?;
+    let receipts = filter_receipts(chunk, receipt_hashes)?;
```

Comment:
Actually this is not what we want because we would like to reproduce the exact same output as if it is applied in production. What we need here is to give transaction/receipt hash as input and then find the right chunk to apply and filter the outcome for that specific transaction or receipt.

Also receipts in the chunk body are outgoing receipts, not incoming receipts. We need incoming receipts for a shard when applying a chunk. You can take a look at https://github.com/near/nearcore/blob/d305aac67b07daf5c168f6b5644fcf58edb1faa9/chain/chain/src/chain.rs#L3521 to see how it is implemented

Reply:
oh good catch... don't know how I missed that. Updated it to try and mimic `save_incoming_receipts_from_block()` by looking in `ColChunkHashesByHeight`

Ok, so I misunderstood what the `--tx` filter here should do. Updated to print out the apply outcomes for any tx or receipt hashes given on the command line. did you mean something like that?

Reply:
btw, I checked that resulting `state_root` is the same as with the `apply` cmd, but the `outcome_root` won't be the same because of the local receipts that hve different hashes.

Reply:
ah also forgot to mention, this last commit still asks for a chunk hash on the command line. I'm trying to think of a good command line interface that would leave it out. With the current `--txs` and `--receipts` flags, you would need to iterate over different chunks (I guess the ones included in `ColChunkHashesByHeight[--target-height]`) and look for one containing your transaction. Then it's a little hairy when you ask what happens if there are two txs given in `--txs` that appear in two different chunks at that height. Do we apply both chunks? or fail? Seems simpler to just ask for a chunk hash to me, since you probably can find the chunk hash you're looking for if you're in a position where you're using this command, but could be wrong about that